src/main/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTask.java

```diff
@@ -146,48 +142,129 @@ private static void checkRateLimit(final MongoSinkTopicConfig config)
-                + '\''
-                + ", details="
-                + bulkWriteError.getDetails()
-                + '}');
+  private class AnalyzedBatchFailedWithBulkWriteException {
+    private final List<SinkRecord> batch;
+    private final MongoBulkWriteException e;
+    private final Map<Integer, Entry<SinkRecord, WriteException>> recordsFailedWithWriteError =
+        new HashMap<>();
+    private final Map<Integer, SinkRecord> recordsFailedWithWriteConcernError = new HashMap<>();
+    private final Map<Integer, SinkRecord> skippedRecords = new HashMap<>();
+    @Nullable private final WriteConcernException writeConcernException;
+    private final WriteSkippedException writeSkippedException = new WriteSkippedException();
+
+    AnalyzedBatchFailedWithBulkWriteException(
+        final List<SinkRecord> batch, final MongoBulkWriteException e) {
+      this.batch = batch;
+      this.e = e;
+      WriteConcernError writeConcernError = e.getWriteConcernError();
+      writeConcernException =
+          writeConcernError == null ? null : new WriteConcernException(writeConcernError);
+      analyze();
+    }
+
+    private void analyze() {
+      List<BulkWriteError> writeErrors = e.getWriteErrors();
+      WriteConcernError writeConcernError = e.getWriteConcernError();
+      if (writeErrors.isEmpty()) {
+        assertNotNull(writeConcernError);
+        for (int i = 0; i < batch.size(); i++) {
+          recordsFailedWithWriteConcernError.put(i, batch.get(i));
+        }
       } else {
-        errorString.add(bulkWriteError.toString());
+        assertTrue(writeErrors.size() == 1);
+        BulkWriteError writeError = writeErrors.get(0);
+        int writeErrorIdx = writeError.getIndex();
+        recordsFailedWithWriteError.put(
+            writeErrorIdx,
+            new SimpleImmutableEntry<>(batch.get(writeErrorIdx), new WriteException(writeError)));
+        for (int i = writeErrorIdx + 1; i < batch.size(); i++) {
+          skippedRecords.put(i, batch.get(i));
+        }
+        if (writeConcernError != null) {
+          for (int i = 0; i < writeErrorIdx; i++) {
+            recordsFailedWithWriteConcernError.put(i, batch.get(i));
+          }
+        }
+      }
+    }
+
+    void log() {
+      if (!recordsFailedWithWriteError.isEmpty()) {
+        recordsFailedWithWriteError.forEach(
+            (i, recordFailedWithWriteError) ->
+                StartedMongoSinkTask.log(
+                    singletonList(recordFailedWithWriteError.getKey()),
+                    recordFailedWithWriteError.getValue()));
+      }
+      if (!recordsFailedWithWriteConcernError.isEmpty()) {
+        StartedMongoSinkTask.log(
+            recordsFailedWithWriteConcernError.values(), assertNotNull(writeConcernException));
+      }
+      if (!skippedRecords.isEmpty()) {
+        StartedMongoSinkTask.log(skippedRecords.values(), writeSkippedException);
+      }
+      LOGGER.error(null, e);
+    }
+
+    /**
```

Comment:
Nit: Spotless rearranges this javadoc (or it does for me locally)

Reply:
It rearranges the doc for me also, but somehow it did not rearrange it previously (I run `./gradlew check -x test -x integrationTest` before committing). Anyway, done.

Reply:
:+1: thanks :)